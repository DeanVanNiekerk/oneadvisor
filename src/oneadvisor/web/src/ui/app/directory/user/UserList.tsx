import { Tag } from "antd";
import React, { Component } from "react";
import { connect, DispatchProp } from "react-redux";

import { getColumnDefinition } from "@/app/table";
import {
    fetchOrganisations,
    Organisation,
    organisationsSelector,
} from "@/state/app/directory/organisations";
import {
    fetchUser,
    fetchUsers,
    receiveUser,
    User,
    UserEdit,
    usersSelector,
} from "@/state/app/directory/users";
import { fetchUsersSimple } from "@/state/app/directory/usersSimple";
import { RootState } from "@/state/rootReducer";
import { Button, getTable, Header } from "@/ui/controls";

import EditUser from "./EditUser";

const Table = getTable<User>();

type Props = {
    users: User[];
    organisations: Organisation[];
    fetching: boolean;
} & DispatchProp;

type State = {
    editVisible: boolean;
};

class UserList extends Component<Props, State> {
    constructor(props: Props) {
        super(props);

        this.state = {
            editVisible: false,
        };
    }

    componentDidMount() {
        if (this.props.users.length === 0) this.loadUsers();
        if (this.props.organisations.length === 0) this.loadOrganisations();
    }

    loadUsers = () => {
        this.props.dispatch(fetchUsers());
        this.props.dispatch(fetchUsersSimple());
    };

    loadOrganisations = () => {
        this.props.dispatch(fetchOrganisations());
    };

    newUser = () => {
        const user: UserEdit = {
            id: null,
            firstName: "",
            lastName: "",
            email: "",
            userName: "",
            branchId: "",
            scope: 1,
            roles: [],
            aliases: [],
            isLocked: false,
            userTypeId: "",
        };

        this.props.dispatch(receiveUser(user));
        this.showEditUser();
    };

    editUser = (id: string) => {
        this.props.dispatch(fetchUser(id));
        this.showEditUser();
    };

    getOrganisationName = (id: string) => {
        const organisation = this.props.organisations.find(u => u.id === id);
        if (organisation) return organisation.name;
    };

    showEditUser = () => {
        this.setState({
            editVisible: true,
        });
    };

    closeEditUser = (cancelled: boolean) => {
        this.setState({
            editVisible: false,
        });
        if (!cancelled) this.loadUsers();
    };

    getColumns = () => {
        const getColumn = getColumnDefinition<User>();
        return [
            getColumn("lastName", "Last Name"),
            getColumn("firstName", "First Name"),
            getColumn("email", "Email"),
            getColumn("organisationName", "Organisation"),
            getColumn("branchName", "Branch"),
            getColumn(
                "emailConfirmed",
                "Activated",
                {},
                {
                    render: (emailConfirmed: boolean) => {
                        return emailConfirmed ? (
                            <Tag color="green">Yes</Tag>
                        ) : (
                            <Tag color="volcano">No</Tag>
                        );
                    },
                }
            ),
            getColumn(
                "isLocked",
                "Locked",
                {},
                {
                    render: (isLocked: boolean) => {
                        return !isLocked ? (
                            <Tag color="green">No</Tag>
                        ) : (
                            <Tag color="volcano">Yes</Tag>
                        );
                    },
                }
            ),
            getColumn(
                "scope",
                "Scope",
                {},
                {
                    render: (scope: number) => {
                        switch (scope) {
                            case 1:
                                return <Tag color="purple">Organisation</Tag>;
                            case 2:
                                return <Tag color="blue">Branch</Tag>;
                            case 3:
                                return <Tag color="magenta">User</Tag>;
                        }
                    },
                }
            ),
        ];
    };

    render() {
        return (
            <>
                <Header
                    icon="team"
                    actions={
                        <Button
                            type="default"
                            icon="plus"
                            onClick={this.newUser}
                            disabled={this.props.fetching}
                            requiredUseCase="dir_edit_users"
                        >
                            New User
                        </Button>
                    }
                >
                    Users
                </Header>
                <Table
                    rowKey="id"
                    columns={this.getColumns()}
                    dataSource={this.props.users}
                    loading={this.props.fetching}
                    onRowClick={user => this.editUser(user.id)}
                />
                <EditUser
                    visible={this.state.editVisible}
                    onClose={this.closeEditUser}
                    organisations={this.props.organisations}
                />
            </>
        );
    }
}

const mapStateToProps = (state: RootState) => {
    const usersState = usersSelector(state);
    const organisationsState = organisationsSelector(state);

    return {
        users: usersState.items,
        organisations: organisationsState.items,
        fetching: usersState.fetching || organisationsState.fetching,
    };
};

export default connect(mapStateToProps)(UserList);
