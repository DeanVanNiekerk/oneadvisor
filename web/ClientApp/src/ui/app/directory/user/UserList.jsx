import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux'
import { Table } from 'reactstrap';

import Loader from 'ui/common/Loader'
import Error from 'ui/common/Error'

import { listSelector } from 'state/app/directory/users/list/selectors'
import { fetchUsers } from 'state/app/directory/users/list/actions'

class UserList extends Component {

  componentDidMount() {
    this.props.dispatch(fetchUsers());
  }

  render() {

    if (this.props.error)
      return <Error />
    if (this.props.fetching)
      return <Loader text="loading users..." />

    return (
      <div>
        <Table>
          <thead>
            <tr>
              <th>Id</th>
              <th>First Name</th>
              <th>Last Name</th>
            </tr>
          </thead>
          <tbody>
                {this.props.users.map(user =>
                    <tr key={user.id}>
                        <td>{user.id}</td>
                        <td>{user.firstName}</td>
                        <td>{user.lastName}</td>
                    </tr>
                )}
            </tbody>
        </Table>
      </div>
    )

  }
}

UserList.propTypes = {
  error: PropTypes.bool.isRequired,
  fetching: PropTypes.bool.isRequired,
  users: PropTypes.array.isRequired,
  dispatch: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  users: listSelector(state).items || [],
  fetching: listSelector(state).fetching,
  error: listSelector(state).error
})

export default connect(mapStateToProps)(UserList)