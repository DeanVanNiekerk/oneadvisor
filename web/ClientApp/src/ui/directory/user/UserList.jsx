import React, { Component } from 'react';
import { connect } from 'react-redux'
import { Table } from 'reactstrap';

import Loader from 'ui/common/Loader'
import Error from 'ui/common/Error'

import { listSelector } from 'state/directory/users/list/selectors'
import { fetchUsers } from 'state/directory/users/list/actions'

class UserList extends Component {

  componentDidMount() {
    this.props.dispatch(fetchUsers());
  }

  render() {

    if (this.props.error)
      return <Error />
    if (this.props.fetching)
      return <Loader text="loading users..." />

    return (
      <div>
        <h3>Users</h3>
        <Table>
          <thead>
            <tr>
              <th>First Name</th>
              <th>Last Name</th>
            </tr>
          </thead>
          <tbody>
                {this.props.users.map(user =>
                    <tr key={user.id}>
                        <td>{user.firstName}</td>
                        <td>{user.lastName}</td>
                    </tr>
                )}
            </tbody>
        </Table>
      </div>
    )

  }
}

const mapStateToProps = state => ({
  users: listSelector(state).items || [],
  fetching: listSelector(state).fetching,
  error: listSelector(state).error
})

export default connect(mapStateToProps)(UserList)