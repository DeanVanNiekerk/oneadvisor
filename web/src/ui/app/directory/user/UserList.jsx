// @flow

import React, { Component } from 'react';

import { connect } from 'react-redux'
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';

import Loader from 'ui/common/Loader'
import Error from 'ui/common/Error'

import type { Dispatch } from 'state/types';
import type { State as RootState } from 'state/rootReducer'
import { listSelector } from 'state/app/directory/users/list/selectors'
import { fetchUsers } from 'state/app/directory/users/list/actions'
import type { User } from 'state/app/directory/users/list/types'

type Props = {
    users: User[],
    fetching: boolean,
    error: boolean,
    dispatch: Dispatch
  };

class UserList extends Component<Props> {

  componentDidMount() {
    this.props.dispatch(fetchUsers());
  }

  render() {

    if (this.props.error)
      return <Error />
    if (this.props.fetching)
      return <Loader text="loading users..." />

    return (
      <Table>
          <TableHead>
            <TableRow>
              <TableCell>Id</TableCell>
              <TableCell>First Name</TableCell>
              <TableCell>Last Name</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
                {this.props.users.map(user =>
                    <TableRow key={user.id}>
                        <TableCell>{user.id}</TableCell>
                        <TableCell>{user.firstName}</TableCell>
                        <TableCell>{user.lastName}</TableCell>
                    </TableRow>
                )}
            </TableBody>
        </Table>
    )

  }
}

const mapStateToProps = (state: RootState) => ({
  users: listSelector(state).items || [],
  fetching: listSelector(state).fetching,
  error: listSelector(state).error
})

export default connect(mapStateToProps)(UserList)