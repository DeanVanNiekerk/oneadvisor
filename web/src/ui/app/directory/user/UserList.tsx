import { Icon, Tag } from 'antd';
import React, { Component } from 'react';
import { connect, DispatchProp } from 'react-redux';

import { getColumn } from '@/app/table';
import { fetchOrganisations, Organisation, organisationsSelector } from '@/state/app/directory/organisations';
import { fetchUser, fetchUsers, receiveUser, User, UserEdit, usersSelector } from '@/state/app/directory/users';
import { fetchUsersSimple } from '@/state/app/directory/usersSimple';
import { RootState } from '@/state/rootReducer';
import { Button, Header, Table } from '@/ui/controls';
import { showMessage } from '@/ui/feedback/notifcation';

import EditUser from './EditUser';

type Props = {
    users: User[];
    organisations: Organisation[];
    fetching: boolean;
} & DispatchProp;

type State = {
    editVisible: boolean;
};

class UserList extends Component<Props, State> {
    constructor(props: Props) {
        super(props);

        this.state = {
            editVisible: false
        };
    }

    componentDidMount() {
        if (this.props.users.length === 0) this.loadUsers();

        if (this.props.organisations.length === 0) this.loadOrganisations();
    }

    loadUsers = () => {
        this.props.dispatch(fetchUsers());
        this.props.dispatch(fetchUsersSimple());
    };

    loadOrganisations = () => {
        this.props.dispatch(fetchOrganisations());
    };

    newUser = () => {
        const user: UserEdit = {
            id: '',
            firstName: '',
            lastName: '',
            email: '',
            login: '',
            branchId: '',
            scope: 1,
            roleIds: [],
            assistantToUserId: '',
            aliases: []
        };

        this.props.dispatch(receiveUser(user));
        this.showEditUser();
    };

    editUser = (id: string) => {
        this.props.dispatch(fetchUser(id));
        this.showEditUser();
    };

    getOrganisationName = (id: string) => {
        const organisation = this.props.organisations.find(u => u.id === id);
        if (organisation) return organisation.name;
    };

    showEditUser = () => {
        this.setState({
            editVisible: true
        });
    };

    closeEditUser = (cancelled: boolean) => {
        this.setState({
            editVisible: false
        });
        if (!cancelled) this.loadUsers();
    };

    getColumns = () => {
        return [
            getColumn('firstName', 'First Name'),
            getColumn('lastName', 'Last Name'),
            getColumn('email', 'Email'),
            getColumn('login', 'Login'),
            getColumn('lastLogin', 'Last Login', { type: 'long-date' }),
            getColumn('organisationId', 'Organisation', {
                render: (organisationId: string) => {
                    return this.getOrganisationName(organisationId);
                }
            }),
            getColumn('status', 'Status', {
                render: (status: string) => {
                    switch (status.toUpperCase()) {
                        case 'STAGED':
                            return <Tag color="cyan">{status}</Tag>;
                        case 'PROVISIONED':
                            return <Tag color="blue">{status}</Tag>;
                        case 'ACTIVE':
                            return <Tag color="green">{status}</Tag>;
                        case 'RECOVERY':
                        case 'LOCKED_OUT':
                        case 'PASSWORD_EXPIRED':
                            return <Tag color="volcano">{status}</Tag>;
                        case 'DEPROVISIONED':
                        case 'SUSPENDED':
                            return <Tag color="red">{status}</Tag>;
                    }
                }
            })
        ];
    };

    render() {
        return (
            <>
                <Header
                    actions={
                        <Button
                            type="default"
                            icon="plus"
                            onClick={this.newUser}
                            disabled={this.props.fetching}
                            requiredUseCase="dir_edit_users"
                        >
                            New User
                        </Button>
                    }
                >
                    Users
                </Header>
                <Table
                    rowKey="id"
                    columns={this.getColumns()}
                    dataSource={this.props.users}
                    loading={this.props.fetching}
                    onRowClick={user => this.editUser(user.id)}
                />
                <EditUser
                    visible={this.state.editVisible}
                    onClose={this.closeEditUser}
                    organisations={this.props.organisations}
                />
            </>
        );
    }
}

const mapStateToProps = (state: RootState) => {
    const usersState = usersSelector(state);
    const organisationsState = organisationsSelector(state);

    return {
        users: usersState.items,
        organisations: organisationsState.items,
        fetching: usersState.fetching || organisationsState.fetching
    };
};

export default connect(mapStateToProps)(UserList);
