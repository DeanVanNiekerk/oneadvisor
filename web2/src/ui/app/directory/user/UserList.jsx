import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux'
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';

import Loader from 'ui/common/Loader'
import Error from 'ui/common/Error'

import { listSelector } from 'state/app/directory/users/list/selectors'
import { fetchUsers } from 'state/app/directory/users/list/actions'

class UserList extends Component {

  componentDidMount() {
    this.props.dispatch(fetchUsers());
  }

  render() {

    if (this.props.error)
      return <Error />
    if (this.props.fetching)
      return <Loader text="loading users..." />

    return (
      <div>
        <Table>
          <TableHead>
            <TableRow>
              <TableCell>Id</TableCell>
              <TableCell>First Name</TableCell>
              <TableCell>Last Name</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
                {this.props.users.map(user =>
                    <TableRow key={user.id}>
                        <TableCell>{user.id}</TableCell>
                        <TableCell>{user.firstName}</TableCell>
                        <TableCell>{user.lastName}</TableCell>
                    </TableRow>
                )}
            </TableBody>
        </Table>
      </div>
    )

  }
}

UserList.propTypes = {
  error: PropTypes.bool.isRequired,
  fetching: PropTypes.bool.isRequired,
  users: PropTypes.array.isRequired,
  dispatch: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  users: listSelector(state).items || [],
  fetching: listSelector(state).fetching,
  error: listSelector(state).error
})

export default connect(mapStateToProps)(UserList)