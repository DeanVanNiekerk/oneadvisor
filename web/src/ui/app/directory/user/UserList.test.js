import React from 'react';
import UserList from './UserList';
import configureMockStore from 'redux-mock-store'
import renderer from 'react-test-renderer';

const mockStore = configureMockStore()

describe('UserList', () => {

    it('renders table', () => {

        const store = mockStore({
            app: {
                directory: {
                    users: {
                        list: {
                            error: false,
                            fetching: false,
                            items: [
                                { id: 1, firstName: 'Jack', lastName: 'Johnson' }
                            ]
                        }
                    }
                }
            }
        })

        store.dispatch = jest.fn();

        const tree = renderer
            .create(<UserList store={store} />)
            .toJSON();

        expect(store.dispatch.mock.calls.length).toBe(1);

        expect(tree).toMatchSnapshot();
    });

    it('renders error', () => {

        const store = mockStore({
            app: {
                directory: {
                    users: {
                        list: {
                            error: true,
                            fetching: false,
                            items: []
                        }
                    }
                }
            }
        })

        store.dispatch = jest.fn();

        const tree = renderer
            .create(<UserList store={store} />)
            .toJSON();

        expect(store.dispatch.mock.calls.length).toBe(1);

        expect(tree).toMatchSnapshot();
    });

    it('renders loader', () => {

        const store = mockStore({
            app: {
                directory: {
                    users: {
                        list: {
                            error: false,
                            fetching: true,
                            items: []
                        }
                    }
                }
            }
        })

        store.dispatch = jest.fn();

        const tree = renderer
            .create(<UserList store={store} />)
            .toJSON();

        expect(store.dispatch.mock.calls.length).toBe(1);

        expect(tree).toMatchSnapshot();
    });

});