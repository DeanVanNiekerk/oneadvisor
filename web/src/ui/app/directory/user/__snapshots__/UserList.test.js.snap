// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList renders error 1`] = `
<div
  className="mt-5 mb-5"
>
  <div
    className="row justify-content-center text-danger"
  >
    <h5>
      An error has occured, please reload the application
    </h5>
  </div>
</div>
`;

exports[`UserList renders loader 1`] = `
<div
  className="MuiGrid-container-17 MuiGrid-direction-xs-column-20 MuiGrid-align-items-xs-center-25 MuiGrid-justify-xs-center-34 sc-bdVaJa cbLjyO"
>
  <div
    className="MuiGrid-item-18"
  >
    <div
      className="MuiCircularProgress-root-114 MuiCircularProgress-colorSecondary-118 MuiCircularProgress-indeterminate-116"
      role="progressbar"
      style={
        Object {
          "height": 40,
          "width": 40,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg-119"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle-120 MuiCircularProgress-circleIndeterminate-122"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </div>
  </div>
  <div
    className="MuiGrid-item-18"
  >
    <p
      className="MuiTypography-root-123 MuiTypography-body1-132 sc-bwzfXH kGFiPV"
    >
      loading users...
    </p>
  </div>
</div>
`;

exports[`UserList renders table 1`] = `
<table
  className="MuiTable-root-1"
>
  <thead
    className="MuiTableHead-root-2"
  >
    <tr
      className="MuiTableRow-root-3 MuiTableRow-head-6"
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root-8 MuiTableCell-head-9"
        scope="col"
      >
        Id
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-8 MuiTableCell-head-9"
        scope="col"
      >
        First Name
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-8 MuiTableCell-head-9"
        scope="col"
      >
        Last Name
      </th>
    </tr>
  </thead>
  <tbody
    className="MuiTableBody-root-16"
  >
    <tr
      className="MuiTableRow-root-3"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-8 MuiTableCell-body-10"
      >
        1
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root-8 MuiTableCell-body-10"
      >
        Jack
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root-8 MuiTableCell-body-10"
      >
        Johnson
      </td>
    </tr>
  </tbody>
</table>
`;
